adr/Registry Refactoring Design.md

```diff
@@ -0,0 +1,144 @@
+
+If **no value** is provided to the `-cp/-configProvider` option, i.e. just `-cp`, and no environment variable override is specified, the default value of `consul.http://localhost:8500` will be used. 
+
+if `-cp/-configProvider` **not used** and no environment variable override is specified the local configuration file is used, as is it now.
+
+The existing `-r/-registry` command line option will be removed from all services. 
+
+### Bootstrap Changes
+
+All services in the edgex-go mono repo use the new common bootstrap functionality.  The plan is to move this code to a go module for the Device Service and App Functions SDKs to also use. The current bootstrap modules `pkg/bootstrap/configuration/registry.go` and `pkg/bootstrap/container/registry.go` will be refactored to use the new `Configuration Client` and be renamed appropriately. New bootstrap modules will be created for using the revised version of `Registry Client` .  The current use of `useRegistry` and `registryClient` for service configuration will be change to appropriate names for using the new `Configuration Client`.  The current use of `useRegistry` and `registryClient` for service registration will be retained for service registration. Call to the new Unregister() API will be added to shutdown code for all services.
+
+### Client Interfaces changes
+
+##### Current Registry Client
+
+This following is the current `Registry Client` Interface
+
+```go
+type Client interface {
+	Register() error
+	HasConfiguration() (bool, error)
+	PutConfigurationToml(configuration *toml.Tree, overwrite bool) error
+	PutConfiguration(configStruct interface{}, overwrite bool) error
+	GetConfiguration(configStruct interface{}) (interface{}, error)
+	WatchForChanges(updateChannel chan<- interface{}, errorChannel chan<- error, 
+                    configuration interface{}, waitKey string)
+	IsAlive() bool
+	ConfigurationValueExists(name string) (bool, error)
+	GetConfigurationValue(name string) ([]byte, error)
+	PutConfigurationValue(name string, value []byte) error
+	GetServiceEndpoint(serviceId string) (types.ServiceEndpoint, error)
+	IsServiceAvailable(serviceId string) error
+}
+```
+
+##### New Configuration Client
+
+This following is the new `Configuration Client` Interface which contains the  **Service Configuration** specific portion from the above current `Registry Client`.
+
+```go
+type Client interface {
+	HasConfiguration() (bool, error)
+	PutConfigurationToml(configuration *toml.Tree, overwrite bool) error
+	PutConfiguration(configStruct interface{}, overwrite bool) error
+	GetConfiguration(configStruct interface{}) (interface{}, error)
+	WatchForChanges(updateChannel chan<- interface{}, errorChannel chan<- error,
+                    configuration interface{}, waitKey string)
+	IsAlive() bool
```

Comment:
IsAlive feels more like registry related?

Reply:
IsAlive() just determines if the Provider is alive, i.e. available. Applicable to Config and Registry